File: enchiridion-go/internal/watch/watch.go
// Package watch is the raw/ watcher — event-driven detection + debounce +
// queue. Ported from `wiki-plugin/scripts/watch_raw.py`.
//
// The `/wiki-watch` skill orchestrates; this is the half it launches in the
// background and polls. Three pieces:
//
//   - [Debouncer] — per-file debounce, pure (injectable clock, no threads, no
//     filesystem) so the timing is testable without real sleeps.
//   - The lock file at `.wiki-knowledge/watch.lock` — one watcher per vault,
//     with stale-lock recovery for a hard-killed predecessor.
//   - The queue file at `.wiki-knowledge/watch-queue.jsonl` — one
//     vault-relative path per line (despite the extension, not JSON). A
//     wake-up signal and nothing more: SKILL.md re-checks the sweep when it
//     needs the reason.
package watch

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// DefaultDebounceSeconds is the per-file settle window.
const DefaultDebounceSeconds = 30.0

// StaleLockSeconds is how long a live-PID lock is trusted before it counts as
// stale.
const StaleLockSeconds = 600

// DefaultPollIntervalSeconds is how often the main loop checks for settled
// files.
const DefaultPollIntervalSeconds = 5.0

// Debouncer tracks the most recent event time per vault-relative file path.
// clock defaults to a monotonic seconds source, injectable so tests can drive
// settling with fake timestamps.
type Debouncer struct {
	debounceSeconds float64
	clock           func() float64
	lastEvent       map[string]float64
}

// NewDebouncer returns a Debouncer with the given settle window. A nil clock
// uses a monotonic source anchored at construction.
func NewDebouncer(debounceSeconds float64, clock func() float64) *Debouncer {
	if clock == nil {
		start := time.Now()
		clock = func() float64 { return time.Since(start).Seconds() }
	}
	return &Debouncer{debounceSeconds: debounceSeconds, clock: clock, lastEvent: map[string]float64{}}
}

// RecordEvent notes an event for rel at the current clock time.
func (d *Debouncer) RecordEvent(rel string) {
	d.lastEvent[rel] = d.clock()
}

// SettledFiles returns, and stops tracking, every file whose debounce window
// has elapsed.
func (d *Debouncer) SettledFiles() []string {
	now := d.clock()
	var settled []string
	for rel, last := range d.lastEvent {
		if now-last >= d.debounceSeconds {
			settled = append(settled, rel)
		}
	}
	for _, rel := range settled {
		delete(d.lastEvent, rel)
	}
	return settled
}

// LastEvent returns the recorded event time for rel, for tests that want to
// assert what the handler recorded.
func (d *Debouncer) LastEvent(rel string) (float64, bool) {
	t, ok := d.lastEvent[rel]
	return t, ok
}

// --- lock file ---------------------------------------------------------------

// WriteLock writes lockPath with the given (or current) PID and timestamp.
func WriteLock(lockPath string, pid int, startedAt time.Time) error {
	if err := os.MkdirAll(filepath.Dir(lockPath), 0o755); err != nil {
		return err
	}
	if pid == 0 {
		pid = os.Getpid()
	}
	if startedAt.IsZero() {
		startedAt = time.Now().UTC()
	}
	payload := map[string]any{
		"pid":        pid,
		"started_at": startedAt.UTC().Format(time.RFC3339Nano),
	}
	data, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	return os.WriteFile(lockPath, data, 0o644)
}

// RemoveLock unlinks lockPath; a no-op when absent.
func RemoveLock(lockPath string) error {
	if err := os.Remove(lockPath); err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}

// lockIsStale reports (isStale, pid) for the lock at lockPath. pid is nil when
// the lock file is unparsable.
//
// An unparsable lock file counts as stale (fails toward proceeding, not toward
// a permanent bail).
func lockIsStale(lockPath string, now time.Time, pidAlive func(int) bool) (bool, *int) {
	data, err := os.ReadFile(lockPath)
	if err != nil {
		return true, nil
	}
	var payload struct {
		PID       int    `json:"pid"`
		StartedAt string `json:"started_at"`
	}
	if err := json.Unmarshal(data, &payload); err != nil {
		return true, nil
	}
	startedAt, err := time.Parse(time.RFC3339Nano, payload.StartedAt)
	if err != nil {
		return true, nil
	}
	pid := payload.PID
	if !pidAlive(pid) {
		return true, &pid
	}
	if now.Sub(startedAt).Seconds() > StaleLockSeconds {
		return true, &pid
	}
	return false, &pid
}

// AcquireLock tries to acquire the watch lock. Returns (acquired, stalePID,
// error).
//
// A live lock (PID alive, within [StaleLockSeconds]) means another watcher is
// running: returns (false, nil, nil), lock untouched. A stale one is removed
// and replaced; the removed lock's PID is returned (nil when the lock file
// was unparsable), so the caller can log the takeover. Check, unlink and
// write all happen under a companion `.mutex` file held with an exclusive
// lock, so two processes racing a stale takeover can't both pass the
// staleness check before either writes.
func AcquireLock(lockPath string, now time.Time, pidAlive func(int) bool) (bool, *int, error) {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	if pidAlive == nil {
		pidAlive = defaultPIDAlive
	}

	if err := os.MkdirAll(filepath.Dir(lockPath), 0o755); err != nil {
		return false, nil, err
	}
	mutexPath := lockPath + ".mutex"
	mutex, err := os.OpenFile(mutexPath, os.O_CREATE|os.O_RDWR, 0o644)
	if err != nil {
		return false, nil, err
	}
	defer mutex.Close()
	if err := exclusiveLock(mutex); err != nil {
		return false, nil, err
	}
	defer exclusiveUnlock(mutex) //nolint:errcheck

	if _, err := os.Stat(lockPath); err == nil {
		stale, pid := lockIsStale(lockPath, now, pidAlive)
		if !stale {
			return false, nil, nil
		}
		if err := os.Remove(lockPath); err != nil {
			return false, nil, err
		}
		if err := WriteLock(lockPath, 0, now); err != nil {
			return false, nil, err
		}
		return true, pid, nil
	}

	return true, nil, WriteLock(lockPath, 0, now)
}

// --- queue file --------------------------------------------------------------

// ReadQueue returns the queue's entries. An absent queue is empty.
//
// Split on "\n", never anything fancier — a path can legitimately contain
// other control bytes.
func ReadQueue(queuePath string) ([]string, error) {
	data, err := os.ReadFile(queuePath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var out []string
	for _, line := range strings.Split(string(data), "\n") {
		if line != "" {
			out = append(out, line)
		}
	}
	return out, nil
}

// withQueueLock runs fn(currentLines) -> newLines under an exclusive file
// lock.
//
// The lock serializes concurrent writers so a read-modify-write can't lose an
// update; writing to a `.tmp` sibling and renaming means a concurrent reader
// never sees a partial write.
func withQueueLock(queuePath string, fn func([]string) []string) error {
	if err := os.MkdirAll(filepath.Dir(queuePath), 0o755); err != nil {
		return err
	}
	writelockPath := queuePath + ".writelock"
	lockFile, err := os.OpenFile(writelockPath, os.O_CREATE|os.O_RDWR, 0o644)
	if err != nil {
		return err
	}
	defer lockFile.Close()
	if err := exclusiveLock(lockFile); err != nil {
		return err
	}
	defer exclusiveUnlock(lockFile) //nolint:errcheck

	existing, err := ReadQueue(queuePath)
	if err != nil {
		return err
	}
	newLines := fn(existing)

	var b strings.Builder
	for _, line := range newLines {
		b.WriteString(line)
		b.WriteString("\n")
	}
	tmpPath := queuePath + ".tmp"
	if err := os.WriteFile(tmpPath, []byte(b.String()), 0o644); err != nil {
		return err
	}
	return os.Rename(tmpPath, queuePath)
}

// AppendQueue appends rel to the queue, unless it's already there
// (idempotent).
func AppendQueue(queuePath, rel string) error {
	return withQueueLock(queuePath, func(lines []string) []string {
		for _, line := range lines {
			if line == rel {
				return lines
			}
		}
		return append(lines, rel)
	})
}

// RemoveFromQueue removes every occurrence of rel from the queue.
func RemoveFromQueue(queuePath, rel string) error {
	return withQueueLock(queuePath, func(lines []string) []string {
		var out []string
		for _, line := range lines {
			if line != rel {
				out = append(out, line)
			}
		}
		return out
	})
}

// --- eligibility check on settle ----------------------------------------------

// CheckAndEnqueue enqueues settledRel iff it's in eligibleRels.
//
// A settled event doesn't mean "ingest this" — an `.ingestignore` match, or a
// file whose back-pointer page is already current, settles too. eligibleRels
// comes from one sweep per poll tick, not per file, so eligibility matches the
// manual sweep exactly.
func CheckAndEnqueue(eligibleRels map[string]bool, settledRel, queuePath string) (bool, error) {
	if !eligibleRels[settledRel] {
		return false, nil
	}
	if err := AppendQueue(queuePath, settledRel); err != nil {
		return false, err
	}
	return true, nil
}

// --- watch paths --------------------------------------------------------------

// Paths is the set of files one watcher run touches.
type Paths struct {
	Root  string
	Lock  string
	Queue string
}

// ForRoot returns the watch paths for a vault root.
func ForRoot(root string) Paths {
	wk := filepath.Join(root, ".wiki-knowledge")
	return Paths{
		Root:  root,
		Lock:  filepath.Join(wk, "watch.lock"),
		Queue: filepath.Join(wk, "watch-queue.jsonl"),
	}
}

// RelForEvent maps one filesystem event path to a vault-relative path, or
// ("", false) when the event should be ignored: a directory, or a path outside
// root. The fsnotify handler calls this per event before debouncing.
func RelForEvent(root, path string) (string, bool) {
	if info, err := os.Stat(path); err != nil || info.IsDir() {
		return "", false
	}
	rel, err := filepath.Rel(root, path)
	if err != nil {
		return "", false
	}
	if rel == ".." || strings.HasPrefix(rel, ".."+string(filepath.Separator)) {
		return "", false
	}
	return filepath.ToSlash(rel), true
}

File: enchiridion-ts/src/watch.ts
/**
 * The raw/ watcher — event-driven detection + debounce + queue.
 *
 * The `/wiki-watch` skill orchestrates; this is the half it launches in the
 * background and polls. Four pieces:
 *
 *   - [runWatch] — the long-lived loop that owns the watcher, the debouncer,
 *     the per-poll-tick sweep, and the queue, with the file-watcher, the
 *     sweep, and the clock injectable so the loop's timing is testable
 *     without chokidar, real signals, or real sleeps. The CLI shrinks to
 *     lock handling and one call to it.
 *   - [Debouncer] — per-file debounce, pure (injectable clock, no threads, no
 *     filesystem) so the timing is testable without real sleeps.
 *   - The lock file at `.wiki-knowledge/watch.lock` — one watcher per vault,
 *     with stale-lock recovery for a hard-killed predecessor.
 *   - The queue file at `.wiki-knowledge/watch-queue.jsonl` — one
 *     vault-relative path per line (despite the extension, not JSON). A
 *     wake-up signal and nothing more: SKILL.md re-checks the sweep when it
 *     needs the reason.
 *
 * Mutual exclusion over the lock/queue files uses an exclusive-create lock
 * file (`.mutex` / `.writelock` created with the `wx` flag, removed on
 * release). The module is pure-JS only (ADR-0017 — a native flock addon
 * would break Bun, the OpenCode runtime), and the atomic `wx` create gives
 * the same cross-process mutual exclusion a blocking flock provides.
 */

import fs from "node:fs";
import path from "node:path";
import { watch as watchRaw } from "chokidar";
import { scan as scanEligible } from "./ingestscan.js";

/** The per-file settle window. */
export const DefaultDebounceSeconds = 30;

/** How long a live-PID lock is trusted before it counts as stale. */
export const StaleLockSeconds = 600;

/** How often the main loop checks for settled files. */
export const DefaultPollIntervalSeconds = 5;

/**
 * Debouncer tracks the most recent event time per vault-relative file path.
 * clock defaults to a monotonic seconds source, injectable so tests can drive
 * settling with fake timestamps.
 */
export class Debouncer {
  private lastEvent = new Map<string, number>();

  constructor(
    private readonly debounceSeconds: number,
    private readonly clock: () => number = defaultClock(),
  ) {}

  /** Notes an event for rel at the current clock time. */
  recordEvent(rel: string): void {
    this.lastEvent.set(rel, this.clock());
  }

  /** Returns, and stops tracking, every file whose debounce window has
   * elapsed. */
  settledFiles(): string[] {
    const now = this.clock();
    const settled: string[] = [];
    for (const [rel, last] of this.lastEvent) {
      if (now - last >= this.debounceSeconds) settled.push(rel);
    }
    for (const rel of settled) this.lastEvent.delete(rel);
    return settled;
  }

  /** Returns the recorded event time for rel, for tests that want to assert
   * what the handler recorded. */
  lastEventTime(rel: string): number | undefined {
    return this.lastEvent.get(rel);
  }
}

function defaultClock(): () => number {
  const start = Date.now();
  return () => (Date.now() - start) / 1000;
}

// --- lock file ---------------------------------------------------------------

/** Writes lockPath with the given (or current) PID and timestamp. */
export function writeLock(
  lockPath: string,
  pid: number,
  startedAt?: Date,
): void {
  fs.mkdirSync(path.dirname(lockPath), { recursive: true, mode: 0o755 });
  if (!pid) pid = process.pid;
  const started = startedAt ? startedAt : new Date();
  const payload = {
    pid,
    started_at: started.toISOString(),
  };
  fs.writeFileSync(lockPath, JSON.stringify(payload), { mode: 0o644 });
}

/** Unlinks lockPath; a no-op when absent. */
export function removeLock(lockPath: string): void {
  try {
    fs.unlinkSync(lockPath);
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code !== "ENOENT") throw err;
  }
}

/** Reports (isStale, pid) for the lock at lockPath. pid is null when the lock
 * file is unparsable.
 *
 * An unparsable lock file counts as stale (fails toward proceeding, not toward
 * a permanent bail). */
function lockIsStale(
  lockPath: string,
  now: Date,
  pidAlive: (pid: number) => boolean,
): { isStale: boolean; pid: number | null } {
  let data: string;
  try {
    data = fs.readFileSync(lockPath, "utf8");
  } catch {
    return { isStale: true, pid: null };
  }
  let payload: { pid?: number; started_at?: string };
  try {
    payload = JSON.parse(data);
  } catch {
    return { isStale: true, pid: null };
  }
  const startedAt = Date.parse(payload.started_at ?? "");
  if (Number.isNaN(startedAt)) return { isStale: true, pid: null };
  const pid = payload.pid ?? 0;
  if (!pidAlive(pid)) return { isStale: true, pid };
  if ((now.getTime() - startedAt) / 1000 > StaleLockSeconds) {
    return { isStale: true, pid };
  }
  return { isStale: false, pid };
}

/** Probes whether pid names a live process via `kill(pid, 0)`: ESRCH means
 * dead, EPERM means alive-but-not-ours. */
export function defaultPIDAlive(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch (err) {
    return (err as NodeJS.ErrnoException).code !== "ESRCH";
  }
}

/**
 * Tries to acquire the watch lock. Returns { acquired, stalePID }.
 *
 * A live lock (PID alive, within [StaleLockSeconds]) means another watcher is
 * running: { acquired: false, stalePID: null }, lock untouched. A stale one is
 * removed and replaced; the removed lock's PID is returned (null when the lock
 * file was unparsable), so the caller can log the takeover. Check, unlink and
 * write all happen under a companion `.mutex` file held exclusively, so two
 * processes racing a stale takeover can't both pass the staleness check before
 * either writes.
 */
export function acquireLock(
  lockPath: string,
  now?: Date,
  pidAlive?: (pid: number) => boolean,
): { acquired: boolean; stalePID: number | null } {
  const nowDate = now ? now : new Date();
  const alive = pidAlive ?? defaultPIDAlive;
  const result: { acquired: boolean; stalePID: number | null } = {
    acquired: false,
    stalePID: null,
  };
  const mutexPath = lockPath + ".mutex";
  withExclusiveLock(mutexPath, () => {
    if (fs.existsSync(lockPath)) {
      const { isStale, pid } = lockIsStale(lockPath, nowDate, alive);
      if (!isStale) {
        result.acquired = false;
        return;
      }
      fs.unlinkSync(lockPath);
      writeLock(lockPath, 0, nowDate);
      result.acquired = true;
      result.stalePID = pid;
      return;
    }
    writeLock(lockPath, 0, nowDate);
    result.acquired = true;
  });
  return result;
}

// --- queue file --------------------------------------------------------------

/** Returns the queue's entries. An absent queue is empty.
 *
 * Split on "\n", never anything fancier — a path can legitimately contain
 * other control bytes. */
export function readQueue(queuePath: string): string[] {
  let data: string;
  try {
    data = fs.readFileSync(queuePath, "utf8");
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") return [];
    throw err;
  }
  const out: string[] = [];
  for (const line of data.split("\n")) {
    if (line !== "") out.push(line);
  }
  return out;
}

/**
 * Runs fn(currentLines) -> newLines under an exclusive lock.
 *
 * The lock serializes concurrent writers so a read-modify-write can't lose an
 * update; writing to a `.tmp` sibling and renaming means a concurrent reader
 * never sees a partial write. */
function withQueueLock(
  queuePath: string,
  fn: (lines: string[]) => string[],
): void {
  fs.mkdirSync(path.dirname(queuePath), { recursive: true, mode: 0o755 });
  const writelockPath = queuePath + ".writelock";
  withExclusiveLock(writelockPath, () => {
    const newLines = fn(readQueue(queuePath));
    let body = "";
    for (const line of newLines) body += line + "\n";
    const tmpPath = queuePath + ".tmp";
    fs.writeFileSync(tmpPath, body, { mode: 0o644 });
    fs.renameSync(tmpPath, queuePath);
  });
}

/** Appends rel to the queue, unless it's already there (idempotent). */
export function appendQueue(queuePath: string, rel: string): void {
  withQueueLock(queuePath, (lines) => {
    for (const line of lines) {
      if (line === rel) return lines;
    }
    return [...lines, rel];
  });
}

/** Removes every occurrence of rel from the queue. */
export function removeFromQueue(queuePath: string, rel: string): void {
  withQueueLock(queuePath, (lines) => lines.filter((l) => l !== rel));
}

// --- eligibility check on settle ----------------------------------------------

/** Enqueues settledRel iff it's in eligibleRels.
 *
 * A settled event doesn't mean "ingest this" — an `.ingestignore` match, or a
 * file whose back-pointer page is already current, settles too. eligibleRels
 * comes from one sweep per poll tick, not per file, so eligibility matches the
 * manual sweep exactly. */
export function checkAndEnqueue(
  eligibleRels: Set<string>,
  settledRel: string,
  queuePath: string,
): boolean {
  if (!eligibleRels.has(settledRel)) return false;
  appendQueue(queuePath, settledRel);
  return true;
}

// --- watch paths --------------------------------------------------------------

/** The set of files one watcher run touches. */
export interface Paths {
  root: string;
  lock: string;
  queue: string;
}

/** Returns the watch paths for a vault root. */
export function forRoot(root: string): Paths {
  const wk = path.join(root, ".wiki-knowledge");
  return {
    root,
    lock: path.join(wk, "watch.lock"),
    queue: path.join(wk, "watch-queue.jsonl"),
  };
}

/** Maps one filesystem event path to a vault-relative path, or null when the
 * event should be ignored: a directory, or a path outside root. */
export function relForEvent(root: string, abs: string): string | null {
  let info: fs.Stats;
  try {
    info = fs.statSync(abs);
  } catch {
    return null;
  }
  if (info.isDirectory()) return null;
  const rel = path.relative(root, abs);
  if (rel === ".." || rel.startsWith(".." + path.sep)) return null;
  return toSlash(rel);
}

// --- exclusive-create lock file (ADR-0017: pure JS, no native addons) ---------

/** Runs critical under an exclusive lock on lockPath, created atomically with
 * the `wx` flag and removed on release.
 *
 * Replaces a blocking flock: an atomic exclusive create gives the same cross-process
 * mutual exclusion without a native addon. When another process holds the
 * lock, blocks (retrying) until it is released. */
function withExclusiveLock(lockPath: string, critical: () => void): void {
  fs.mkdirSync(path.dirname(lockPath), { recursive: true, mode: 0o755 });
  let fd: number;
  for (;;) {
    try {
      fd = fs.openSync(lockPath, "wx");
      break;
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== "EEXIST") throw err;
      sleep(5);
    }
  }
  let criticalErr: unknown = null;
  try {
    critical();
  } catch (err) {
    criticalErr = err;
  } finally {
    fs.closeSync(fd);
    try {
      fs.unlinkSync(lockPath);
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== "ENOENT") {
        if (criticalErr === null) criticalErr = err;
      }
    }
  }
  if (criticalErr !== null) throw criticalErr;
}

function sleep(ms: number): void {
  Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, ms);
}

function toSlash(p: string): string {
  return p.split(path.sep).join("/");
}

// --- the watch loop -----------------------------------------------------------

/** The minimal file-watcher surface [runWatch] drives. Chokidar's FSWatcher
 * satisfies it structurally. */
export interface Watcher {
  /** Registers a filesystem-event handler. The loop consumes only the "all"
   * event, keyed by (eventName, path). */
  on(event: "all", cb: (eventName: string, p: string) => void): void;
  /** Registers the log-and-keep-watching error handler. */
  on(event: "error", cb: (err: unknown) => void): void;
  /** Registers the ready handler, which fires the "watching …" banner. */
  on(event: "ready", cb: () => void): void;
  /** Stops watching; resolves once fully closed. */
  close(): Promise<void>;
}

/** One eligibility sweep, run per poll tick when files have settled. Returns
 * the vault-relative paths the sweep offers. */
export type Sweep = () => Promise<Set<string>>;

/** Schedules `tick` every `ms`; the returned function cancels it. The default
 * uses setInterval/clearInterval; tests inject a scheduler that stores the
 * tick for manual driving, so no real sleeps are needed. */
export type Scheduler = (tick: () => void, ms: number) => () => void;

/** The injectable seams [runWatch] composes. Every field defaults to the
 * production implementation; tests inject fakes for all of them. */
export interface WatchOptions {
  /** Per-file settle window, seconds (default [DefaultDebounceSeconds]). */
  debounceSeconds?: number;
  /** How often to check for settled files, seconds (default
   * [DefaultPollIntervalSeconds]). */
  pollIntervalSeconds?: number;
  /** Creates the file watcher over the raw/ root (default: chokidar). */
  makeWatcher?: (rawRoot: string) => Watcher;
  /** One eligibility sweep per poll tick (default: the real ingest-scan over
   * paths.root, matching the manual sweep exactly). */
  sweep?: Sweep;
  /** Monotonic-seconds clock for the debouncer (default: real time). */
  clock?: () => number;
  /** Registers a signal handler (default: process.on). */
  onSignal?: (signal: "SIGINT" | "SIGTERM", cb: () => void) => void;
  /** Removes a signal handler (default: process.removeListener). */
  offSignal?: (signal: "SIGINT" | "SIGTERM", cb: () => void) => void;
  /** Schedules the per-poll-tick sweep (default: setInterval). */
  schedule?: Scheduler;
  /** The pid printed in the "watching …" banner (default: process.pid). */
  pid?: number;
  /** Emits the loop's user-facing lines (default: console.log). */
  log?: (line: string) => void;
}

/**
 * runWatch runs the long-lived watch loop: a file watcher over root/raw/
 * records events into the debouncer; on each poll tick settled files are
 * swept for eligibility and enqueued. SIGINT/SIGTERM logs "watcher stopped",
 * cancels the poll, closes the watcher, removes the lock, and resolves the
 * returned promise.
 *
 * Creates root/raw/ if missing. The watcher, sweep, clock, signal handling,
 * scheduler, and log are injectable, so the loop is testable with no
 * chokidar, no real signals, and no real sleeps.
 */
export function runWatch(
  paths: Paths,
  options: WatchOptions = {},
): Promise<void> {
  const debounceSeconds = options.debounceSeconds ?? DefaultDebounceSeconds;
  const pollIntervalSeconds =
    options.pollIntervalSeconds ?? DefaultPollIntervalSeconds;
  const rawRoot = path.join(paths.root, "raw");
  fs.mkdirSync(rawRoot, { recursive: true, mode: 0o755 });

  const watcher = (options.makeWatcher ?? defaultWatcher)(rawRoot);
  const debouncer = new Debouncer(
    debounceSeconds,
    options.clock ?? defaultClock(),
  );
  const sweep = options.sweep ?? defaultSweep(paths.root);
  const schedule = options.schedule ?? defaultSchedule;
  const onSignal = options.onSignal ?? defaultOnSignal;
  const offSignal = options.offSignal ?? defaultOffSignal;
  const log = options.log ?? console.log;
  const pid = options.pid ?? process.pid;

  return new Promise<void>((resolve) => {
    let cancel = (): void => {};
    let stopped = false;
    const stop = (): void => {
      if (stopped) return;
      stopped = true;
      log("watcher stopped");
      cancel();
      offSignal("SIGINT", stop);
      offSignal("SIGTERM", stop);
      void watcher.close();
      removeLock(paths.lock);
      resolve();
    };
    onSignal("SIGINT", stop);
    onSignal("SIGTERM", stop);

    watcher.on("all", (_eventName: string, p: string) => {
      const rel = relForEvent(paths.root, p);
      if (rel !== null) debouncer.recordEvent(rel);
    });
    watcher.on("error", (_err: unknown) => {
      // Log-and-keep-watching; a transient read error isn't fatal.
    });
    watcher.on("ready", () => {
      log(`watching ${rawRoot} (debounce=${debounceSeconds}s, pid=${pid})`);
    });

    cancel = schedule(() => {
      const settled = debouncer.settledFiles();
      if (settled.length === 0) return;
      sweep()
        .then((eligible) => {
          for (const rel of settled) {
            const queued = checkAndEnqueue(eligible, rel, paths.queue);
            if (queued) log(`queued ${rel}`);
          }
        })
        .catch((err) => {
          log(`error scanning raw/: ${(err as Error).message}`);
        });
    }, pollIntervalSeconds * 1000);
  });
}

function defaultWatcher(rawRoot: string): Watcher {
  return watchRaw(rawRoot, { ignoreInitial: true });
}

function defaultSweep(root: string): Sweep {
  return async (): Promise<Set<string>> => {
    const result = await scanEligible(root, "", null);
    return new Set(result.eligible.map((c) => c.rawRel));
  };
}

function defaultSchedule(tick: () => void, ms: number): () => void {
  const id = setInterval(tick, ms);
  return () => clearInterval(id);
}

function defaultOnSignal(signal: "SIGINT" | "SIGTERM", cb: () => void): void {
  process.on(signal, cb);
}

function defaultOffSignal(signal: "SIGINT" | "SIGTERM", cb: () => void): void {
  process.removeListener(signal, cb);
}
